src/client/datascience/interactive-window/nativeInteractiveWindow.ts

```diff
@@ -130,20 +136,31 @@ export class NativeInteractiveWindow implements IInteractiveWindowLoadable {
             this._submitters.push(owner);
         }
 
+        this.initialControllerSelected = createDeferred<void>();
+
+        // Immediately try to find a selected controller for our NotebookDocument,
+        // as it's possible that a selection event fired before our ctor was able to run
+        const controller = this.notebookControllerManager.getSelectedNotebookController(this.notebookDocument);
+        if (controller !== undefined) {
+            this.registerKernel(this.notebookDocument, controller);
+        }
+
+        // Ensure we hear about any controller changes so we can update our cache accordingly
         this.notebookControllerManager.onNotebookControllerSelected(
             (e: { notebook: NotebookDocument; controller: VSCodeNotebookController }) => {
-                if (this._notebookUri !== undefined && e.notebook.uri.toString() !== this._notebookUri.toString()) {
+                if (e.notebook.uri.toString() !== this.notebookDocument.uri.toString()) {
                     return;
                 }
 
                 // Clear cached kernel when the selected controller for this document changes
-                const controllerChangeListener = e.controller.controller.onDidChangeSelectedNotebooks(
+                const controllerChangeListener = (
+                    this.notebookController || e.controller
```

Comment:
Why not just leave this as `e.controller`?
Why give preference to `this.notebookController`?